### PR TITLE
Add basic flow functions and creation methods

### PR DESCRIPTION
**Pull request by uini223**

### CODE REVIEW

File: flows/src/main/java/com/softwaremill/jox/ThrowingConsumer.java [hunk 1]
```diff
@@ -0,0 +1,5 @@
+package com.softwaremill.jox;
+
+public interface ThrowingConsumer<T> {
+    void accept(T t) throws Throwable;
```
Comment (adamw) [inline flows/src/main/java/com/softwaremill/jox/ThrowingConsumer.java:4]: wouldn't `... throws Exception` be enough? (here and in other places)

Comment (uini223) [reply flows/src/main/java/com/softwaremill/jox/ThrowingConsumer.java]: Yes, it will be enough. Changed

File: flows/pom.xml [hunk 2]
```diff
@@ -0,0 +1,38 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (adamw) [inline flows/pom.xml:1]: the main pom should also list `flows` as a module?

File: flows/src/main/java/com/softwaremill/jox/Flow.java [hunk 3]
```diff
@@ -0,0 +1,320 @@
+package com.softwaremill.jox;
```
Comment (adamw) [inline flows/src/main/java/com/softwaremill/jox/Flow.java:1]: maybe let's put this in a dedicated package, so that there's a package-per-module

File: flows/src/test/java/com/softwaremill/jox/FlowsTest.java [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+
+class FlowsTest {
+
+    @Test
+    void shouldBeEmpty() throws Exception {
+        assertTrue(Flows.empty().runToList().isEmpty());
+    }
+
+    @Test
+    @WithUnsupervisedScope
```
Comment (adamw) [inline flows/src/test/java/com/softwaremill/jox/FlowsTest.java:27]: does this do anything other than calling `Scopes.unsupervised(...)`? If so, it only saves one line, so maybe not worth it?

**Comment by adamw** [comment]
I think this looks relatively complete? If so, mark as ready :)

File: flows/src/test/java/com/softwaremill/jox/FlowsTest.java (hunk 4 above)
Comment (uini223) [reply flows/src/test/java/com/softwaremill/jox/FlowsTest.java:27]: Yes, it only calls Scopes.unsupervised and passes scope as argument.

IMO it makes the tests a little bit cleaner, there is no need to wrap the whole test in lambda and there is no ugly return null at the end. 

So 
```java
    @Test
    void shouldCreateFlowFromFork() throws Exception {
        Scopes.unsupervised(scope -> {
            Fork<Integer> f = scope.forkUnsupervised(() -> 1);
            Flow<Integer> c = Flows.fromFork(f);
            assertEquals(List.of(2), c.runToList());
            return null;
        });
    }
``` 
vs 

```java
    @Test
    @WithUnsupervisedScope
    void shouldCreateFlowFromFork(UnsupervisedScope scope) throws Exception {
        Fork<Integer> f = scope.forkUnsupervised(() -> 1);
        Flow<Integer> c = Flows.fromFork(f);
        assertEquals(List.of(1), c.runToList());
    }
```

File: flows/src/main/java/com/softwaremill/jox/Flow.java (hunk 3 above)
Comment (uini223) [reply flows/src/main/java/com/softwaremill/jox/Flow.java]: Done

File: flows/pom.xml (hunk 2 above)
Comment (uini223) [reply flows/pom.xml:1]: Yes, it should. Added.

File: flows/src/test/java/com/softwaremill/jox/FlowsTest.java (hunk 4 above)
Comment (adamw) [reply flows/src/test/java/com/softwaremill/jox/FlowsTest.java]: Yeah the `return null` is indeed ugly, but I think I'd opt for explicit `Scopes.unsupervised` calls, as that's how the flows/scopes should be used. That is, the scope is not some context that "normally" will be available from another place, but rather the primary means of usage.

Comment (uini223) [reply flows/src/test/java/com/softwaremill/jox/FlowsTest.java]: Ok, makes sense, will change it then

Comment (uini223) [reply flows/src/test/java/com/softwaremill/jox/FlowsTest.java]: done

File: flows/src/main/java/com/softwaremill/jox/flows/Flow.java [hunk 5]
```diff
@@ -0,0 +1,322 @@
...
+        runForeach(result::add);
+        return result;
+    }
+
+    /** The flow is run in the background, and each emitted element is sent to a newly created channel, which is then returned as the result
+     * of this method.
+     * <p>
+     * By default, buffer capacity is unlimited.
+     * <p>
+     * Blocks until the flow completes.
```
Comment (adamw) [inline flows/src/main/java/com/softwaremill/jox/flows/Flow.java:65]: this comment is incorrect/lacking:
1. the channel is rendezvous (capacity 0), not unlimited. Rendezvous channels might mean poor capacity, so I think we should instead by default use a 16-buffered channel (as in Ox). Question is, how to provide this data. Explicit parameters are ... explicit, and not that nice. Scope-local value, maybe? This could be used in other places a swell
2. the method doesn't block - that's also incorrect in Ox's docs, I'll fix it in a second
3. there's no mention on why you need a supervision scope - and I think this might not be obvious to users

Comment (adamw) [inline flows/src/main/java/com/softwaremill/jox/flows/Flow.java:65]: this comment is incorrect/lacking:
1. the channel is rendezvous (capacity 0), not unlimited. Rendezvous channels might mean poor capacity, so I think we should instead by default use a 16-buffered channel (as in Ox). Question is, how to provide this data. Explicit parameters are ... explicit, and not that nice. Scope-local value, maybe? This could be used in other places a swell
2. the method doesn't block - that's also incorrect in Ox's docs, I'll fix it in a second
3. there's no mention on why you need a supervision scope - and I think this might not be obvious to users

**Review by adamw — APPROVED** [review]

File: flows/src/main/java/com/softwaremill/jox/flows/Flow.java (hunk 5 above)
Comment (uini223) [reply flows/src/main/java/com/softwaremill/jox/flows/Flow.java]: 1. In the next MR, I've added overloaded method with buffer capacity as additional parametr, but there are more methods with similar problems.
I've added ScopedValue in Channel together with default size. 
I'll also add overloaded method in second MR to keep flexibility. 

We might want to consider adding Value Object for buffer capacity to keep things like unlimited buffer, rendevous size, default size, run with scoped etc.

2. Fixed
3. Added, please review if is understandable

**Comment by adamw** [comment]
Thanks, let's continue on other PRs :)
